src/compiler/checker.ts

```diff
@@ -9871,31 +9871,40 @@ namespace ts {
             }
 
             // Functions with with an explicitly specified 'void' or 'any' return type don't need any return expressions.

```

Comment:
'explicitly specified' reflects what is checked in the `if` below - if function has explicit type annotation and it is `void` / `any`